File: RightSideViewOfBinaryTree.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class RightSideViewOfBinaryTree {
    List<Integer> result;
    public List<Integer> rightSideView(TreeNode root) {
        result=new ArrayList<>();
        
        //dfs(root,0);
        bfs(root);
        return result;
    }
     // Time Complexity : O(n)
    // Space Complexity : O(h)
    // Did this code successfully run on Leetcode : yes
    


    public void dfs(TreeNode root,int level)
    {
        if(root==null)return;
        if(result.size()==level)
        {
            result.add(root.val);
            
        }
        else {
            result.set(level,root.val);
        }
        dfs(root.left,level+1);
        dfs(root.right,level+1);
    }
    // Time Complexity : O(n)
    // Space Complexity : O(n)
    // Did this code successfully run on Leetcode : yes

    public void bfs(TreeNode root)
    {
       if(root==null)
           return;
        Queue<TreeNode> queue=new LinkedList<>();
        queue.add(root);
        //because you're getting last node of each level so maintaining level is important
        while(!queue.isEmpty())
        {
            int size=queue.size();
            for(int i=0;i<size;i++)
            {
                TreeNode node=queue.poll();
                if(i==size-1)
                    result.add(node.val);
                if(node.left!=null)
                    queue.add(node.left);
                if(node.right!=null)
                    queue.add(node.right);
                
            }
        }
    }
}

File: rightSideViewOfBinaryTree.py
# // Time Complexity :O(N) we would have to visit each node for BFS and DFS
# // Space Complexity : O(D) --> leaf nodes for BFS, O(H)--> height for DFS
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No


# // Your code here along with comments explaining your approach
## In BFS solution we addpend each level to our queue but we only add the last element in the level to our result as 
## that will be the rightmost element.
class solution:
    def bfs(self, root):
        res = []

        queue=collections.deque()

        queue.append(root)

        while queue:

            qsize = len(queue)
            for i in range(qsize):
                curr = queue.pop()

                if i == qsize-1:
                    res.append(curr)
                
                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
        
        return res
## In DFS however we need to go deep in the branch, so in the first way what we do is instead of branching to left first
## we branch to right and if len of res == level of our node we append it . This way we make sure only right is going.
## in case we have to do left first then we amend the value in a particular level as we traverse through the tree.
    def dfs(self, root):
        self.res = []
        self.helper(root, 0)
        return self.res
    
    def helper(self, root, level):

        ##base
        if root == None:
            return 
        ##logic

        if len(self.res) == level:
            self.res.append(root.val)

        self.helper(root.right, level+1)

        self.helper(root.left, level+1)
    
    def helper_left_first(self,root, level):

        ## base
        if root == None:
            return

        ##logic
        if len(self.res) == level:
            self.res.append(root.val)
        else:
            self.res[level]= root.val
        self.helper_left_first(root.left, level+1)

        self.helper_left_first(root.right, level+1)
